crates/astria-sequencer-relayer/src/relayer.rs

```diff
@@ -97,62 +92,85 @@ impl Relayer {
-        }
-
-        info!(height = ?height, tx_count = resp.block.data.len(), "got block from sequencer");
-        new_state.current_sequencer_height.replace(height);
-
-        if resp.block.header.proposer_address.as_ref() != self.validator.address.as_ref() {
-            let proposer_address = resp.block.header.proposer_address;
-            info!(
-                %proposer_address,
-                validator_address = %self.validator.address,
-                "ignoring block: proposer address is not ours",
+    fn convert_block_response_to_sequencer_block_data(
+        &self,
+        res: BlockResponse,
+        current_height: Option<u64>,
+    ) -> eyre::Result<SequencerBlockData> {
+        if let Some(current_height) = current_height {
+            ensure!(
+                res.block.header.height.value() > current_height,
+                "sequencer block response had height below current height tracked in relayer"
             );
-            return Ok(new_state);
         }
+        ensure!(
+            res.block.header.proposer_address == self.validator.address,
+            "proposer recorded in sequencer block does not match internal validator"
+        );
+        let sequencer_block_data = SequencerBlockData::from_tendermint_block(res.block)
+            .wrap_err("failed converting sequencer block response to sequencer block data")?;
+        Ok(sequencer_block_data)
+    }
 
-        let sequencer_block = match SequencerBlockData::from_tendermint_block(resp.block) {
-            Ok(block) => block,
-            Err(e) => {
-                warn!(error = ?e, "failed to convert block to DA block");
-                return Ok(new_state);
+    #[instrument(skip_all)]
+    async fn submit_blocks(&self, block_responses: Vec<BlockResponse>) -> State {
+        let mut new_state = (*self.state_tx.borrow()).clone();
+        let mut all_converted_blocks = Vec::with_capacity(block_responses.len());
+        for res in block_responses {
+            match self.convert_block_response_to_sequencer_block_data(
+                res,
+                new_state.current_sequencer_height,
+            ) {
+                Ok(converted) => {
+                    if let Err(error) = self.block_tx.send(converted.clone()) {
+                        warn!(?error, "failed sending sequencer block data to gossip task");
+                    }
+                    all_converted_blocks.push(converted)
+                }
+                Err(error) => {
+                    // TODO: better event field, maybe with a way to identify the block?
+                    warn!(?error, "dropping block response");
+                }
             }
+        }
+        // get the max sequencer height from the valid blocks; the result of this op is `None` if
+        // the vector is empty.
+        let Some(max_sequencer_height) = all_converted_blocks
+            .iter()
+            .map(|block| block.header.height.value())
+            .max()
```

Comment:
this, combined with the `get_latest_block` for polling means blocks can potentially be missed if the polling time isn't fast enough.

Reply:
Wouldn't we always miss blocks if we don't poll fast enough? Looking at the sequencer client there is I think no way to get a block at a particular height.

IMO we should probably get rid of the http client in favor of the websocket one and subscribe to [`EventType::NewBlock`](https://docs.rs/tendermint-rpc/0.32.0/tendermint_rpc/query/enum.EventType.html#variant.NewBlock)

There is still the danger of missing events, but at least we don't have to keep track internally of what the latest block was.

Hm, I am not sure if tower-abci supports websockets tho. :(

Reply:
there is a function to get block by height: https://github.com/astriaorg/astria/blob/main/crates/astria-sequencer-client/src/client.rs#L71

yeah i agree, pushing would be preferred over polling

Reply:
https://github.com/astriaorg/astria/issues/181